pkg/porter/lifecycle.go

```diff
@@ -26,22 +30,199 @@ type BundleAction interface {
+	if err := o.parseParams(); err != nil {
+		return err
+	}
+
+	o.defaultDriver(p)
+	if err := o.validateDriver(p.Context); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// LoadParameters validates and resolves the parameters and sets. It must be
+// called after porter has loaded the bundle definition.
+func (o *BundleExecutionOptions) LoadParameters(ctx context.Context, p *Porter, bun cnab.ExtendedBundle) error {
+	// This is called in multiple code paths, so exit early if
+	// we have already loaded the parameters into combinedParameters
+	if o.combinedParameters != nil {
+		return nil
+	}
+
+	err := o.parseParams()
+	if err != nil {
+		return err
+	}
+
+	err = o.parseParamSets(ctx, p, bun)
+	if err != nil {
+		return err
+	}
+
+	o.combinedParameters = o.combineParameters(p.Context)
+
+	return nil
+}
+
+// parsedParams parses the variable assignments in Params.
+func (o *BundleExecutionOptions) parseParams() error {
+	p, err := storage.ParseVariableAssignments(o.Params)
+	if err != nil {
+		return err
+	}
+
+	o.parsedParams = p
+	return nil
+}
+
+func (o *BundleExecutionOptions) populateInternalParameterSet(ctx context.Context, p *Porter, bun cnab.ExtendedBundle, i *storage.Installation) error {
+	strategies := make([]secrets.Strategy, 0, len(o.parsedParams))
+	for name, value := range o.parsedParams {
+		strategies = append(strategies, storage.ValueStrategy(name, value))
+	}
+
+	strategies, err := p.Sanitizer.CleanParameters(ctx, strategies, bun, i.ID)
+	if err != nil {
+		return err
+	}
+
+	if len(strategies) == 0 {
+		// if no override is specified, clear out the old parameters on the
+		// installation record
+		i.Parameters.Parameters = nil
+		return nil
+	}
+
+	i.Parameters = i.NewInternalParameterSet(strategies...)
+
+	return nil
+}
+
+// parseParamSets parses the variable assignments in ParameterSets.
+func (o *BundleExecutionOptions) parseParamSets(ctx context.Context, p *Porter, bun cnab.ExtendedBundle) error {
+	if len(o.ParameterSets) > 0 {
+		parsed, err := p.loadParameterSets(ctx, bun, o.Namespace, o.ParameterSets)
+		if err != nil {
+			return fmt.Errorf("unable to process provided parameter sets: %w", err)
+		}
+		o.parsedParamSets = parsed
+	}
+	return nil
+}
+
+// Combine the parameters into a single map
+// The params set on the command line take precedence over the params set in
+// parameter set files
+// Anything set multiple times, is decided by "last one set wins"
+func (o *BundleExecutionOptions) combineParameters(c *portercontext.Context) map[string]string {
+	final := make(map[string]string)
+
+	for k, v := range o.parsedParamSets {
+		final[k] = v
+	}
+
+	for k, v := range o.parsedParams {
+		final[k] = v
+	}
+
+	//
+	// Default the porter-debug param to --debug
+	//
+	if c.Debug {
+		final["porter-debug"] = "true"
+	}
+
+	return final
+}
+
+// defaultDriver supplies the default driver if none is specified
+func (o *BundleExecutionOptions) defaultDriver(p *Porter) {
+	if o.Driver == "" {
```

Comment:
The comments are great! However, I'm still not sure why we need to check `o.Driver == ""` twice here?

Reply:
oops! 🤦‍♀️ That was a copy/paste error when I moved the code from cnab.go to lifecycle.go. I'll get that fixed!